src/libcore/iter.rs

```diff
@@ -369,6 +369,25 @@ pub trait Iterator {
     /// `None` here means that either there is no known upper bound, or the
     /// upper bound is larger than `usize`.
     ///
+    /// # Safety
```

Comment:
@bluss I meant client code must not trust this function to perform unsafe operations.

For example, `Vec::size()` can be trused, that vec contains exactly `size` elements. You can safely write code like this:

```
ptr = malloc(vec.size() * elem_size);
for item in vec {
   *ptr = item;
   ptr += elem_size;
}
```

But you should not do similar things with any `Iterator`, because `size_hint` may be incorrect.

You could do it with [`TrustedIterator`](https://github.com/bluss/rust-itertools/blob/master/src/ziptrusted.rs#L7) from itertools, which is unsafe trait, but not with `Iterator`.

So yes, it is about safety.